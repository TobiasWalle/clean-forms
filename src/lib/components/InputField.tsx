import * as React from 'react';
import { createField } from '../hocs';
import { getInputProps } from '../utils/getInputProps';

export const InputField = createField<string, JSX.IntrinsicElements['input']>(
  ({ input: fieldInputProps, custom }) => {
    return <input {...getInputProps(fieldInputProps, custom)} />;
  }
);
