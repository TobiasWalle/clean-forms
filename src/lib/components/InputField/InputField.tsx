import * as React from 'react';
import { createField } from '../../hocs';
import { FieldPropsWithoutRender, FieldRenderFunction, } from '../Field/Field';

export type InputValue = number | string;

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string;
}

export const renderInput: FieldRenderFunction<InputValue, InputProps> = ({ input: { name, value, onChange }, custom }) => {
  const label = custom && custom.label;
  return <label>
    <div>{label}</div>
    <input {...custom} name={name} value={value} onChange={event => onChange(event.target.value)}/>
  </label>;
};

export const InputField = createField<InputValue, InputProps>(renderInput);
