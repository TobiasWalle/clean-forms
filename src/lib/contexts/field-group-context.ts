import * as React from 'react';
import { Path } from '../utils';

export interface FieldGroupContextValue {
  namespace?: Path;
  path?: Path;
}

export const FieldGroupContext = React.createContext<FieldGroupContextValue>({});
