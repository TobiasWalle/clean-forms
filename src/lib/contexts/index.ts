export * from './fieldContext';
export * from './fieldArrayContext';
