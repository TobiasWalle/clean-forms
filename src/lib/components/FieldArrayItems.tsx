import * as React from 'react';
import { memo, useCallback, useMemo } from 'react';
import { useFieldArrayContext } from '../contexts/field-array-context';
import { FieldContext, FieldContextProvider, FieldContextValue, useFieldContext } from '../contexts/field-context';
import { DELETE } from '../utils';

export type SetArray<Item> = (newArray: Item[]) => void;

export interface InnerFieldArrayItemProps<Item> {
  remove: () => void;
  item: Item;
  index: number;
  setArray: SetArray<Item>;
}

export type GetKey<Item> = (item: Item, index: number) => any;

export type FieldArrayItemsRender<Item> = (props: InnerFieldArrayItemProps<Item>) => React.ReactElement;

export interface FieldArrayItemsProps<Item> {
  render: FieldArrayItemsRender<Item>;
}

function _FieldArrayItems<Item = any>(props: FieldArrayItemsProps<Item>) {
  const { getKey } = useFieldArrayContext();
  const {
    value: array,
    setValue: setArray,
  } = useFieldContext<Item[]>();
  const { render } = props;

  const elements = useMemo(() => {
    return (
      array.map((item, index) => <FieldArrayItem<Item>
        setArray={setArray}
        key={getKey(item, index)}
        item={item}
        index={index}
        render={render}
      />)
    );
  }, [array, getKey, render, setArray]);

  return (
    <>
      {elements}
    </>
  );
}

export type FieldArrayItems = typeof _FieldArrayItems;
export const FieldArrayItems: FieldArrayItems = memo(_FieldArrayItems) as any;

interface FieldArrayItemProps<Item> {
  index: number;
  item: Item;
  setArray: SetArray<Item>;
  render: FieldArrayItemsRender<Item>;
}

function _FieldArrayItem<Item>({
  index,
  item,
  setArray,
  render,
}: FieldArrayItemProps<Item>) {
  const { getKey } = useFieldArrayContext();
  const relativeFieldPath = String(getKey(item, index));
  const relativeModelPath = String(index);
  const renderItem = useCallback(
    (fieldContext: FieldContextValue<Item>) => (
      render({
        remove: () => fieldContext.setValue(DELETE as any),
        setArray,
        index,
        item,
      })
    ),
    [render, setArray, index, item],
  );

  return (
    <FieldContextProvider
      relativeFieldPath={relativeFieldPath}
      relativeModelPath={relativeModelPath}
    >
      <FieldContext.Consumer>
        {renderItem as any}
      </FieldContext.Consumer>
    </FieldContextProvider>
  );
}

const FieldArrayItem: typeof _FieldArrayItem = memo(_FieldArrayItem) as any;
