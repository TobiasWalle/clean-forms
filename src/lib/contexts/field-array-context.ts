import * as React from 'react';
import { GetKey } from '../components';

export interface FieldArrayContextValue {
  getKey: GetKey<any>;
}

export const defaultFieldArrayContextValue: FieldArrayContextValue = {
  getKey: (item, index) => `${index}`
};

export const FieldArrayContext = React.createContext<FieldArrayContextValue>(defaultFieldArrayContextValue);
