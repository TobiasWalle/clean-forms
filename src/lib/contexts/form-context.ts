import { useContext } from 'react';
import * as React from 'react';
import { FormApi } from '../api';
import { GetKey } from '../components';
import { assertNotNull, Path } from '../utils';

export type OnFieldMount = (id: string) => void;

export type OnFieldUnmount = (id: string) => void;

export type OnFieldFocus = (id: string) => void;

export type OnFieldBlur = (id: string) => void;

export type OnFieldChange<Model> = (id: string, path: Path, value: any) => void;

export type SetArrayGetKey = (id: Path, getKey: GetKey<any>) => void;

export interface FormContextValue<Model> {
  form: FormApi<Model>;
  onFieldMount: OnFieldMount;
  onFieldUnmount: OnFieldUnmount;
  onFieldFocus: OnFieldFocus;
  onFieldBlur: OnFieldBlur;
  onFieldChange: OnFieldChange<Model>;
  setArrayGetKey: SetArrayGetKey;
}

export const FormContext = React.createContext<FormContextValue<any> | null>(null);

export function useFormContext<T>(): FormContextValue<T> {
  return assertNotNull(
    useContext(FormContext),
    'This component needs to be used inside a FormContext'
  );
}
