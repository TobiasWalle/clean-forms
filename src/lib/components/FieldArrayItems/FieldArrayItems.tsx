import * as React from 'react';
import { FieldGroup } from '../';
import { FieldArrayContext, FieldArrayContextValue } from '../../contexts/field-array-context';
import { FieldGroupContext, FieldGroupContextValue } from '../../contexts/field-group-context';
import { FormContext, FormContextValue } from '../../contexts/form-context';
import { assertNotNull, DELETE, selectDeep } from '../../utils';
import { createPath, Path } from '../../utils';
import { isShallowEqual } from '../../utils/isShallowEqual';

export interface InnerFieldArrayItemProps<Item> {
  remove: () => void;
  item: Item;
  index: number;
  setArray: (newArray: Item[]) => void;
}

export type GetKey<Item> = (item: Item, index: number) => any;

export type FieldArrayItemsRender<Item> = (props: InnerFieldArrayItemProps<Item>) => React.ReactNode;

export interface FieldArrayItemsProps<Item> {
  render: FieldArrayItemsRender<Item>;
}

export class FieldArrayItems<Item = any> extends React.PureComponent<FieldArrayItemsProps<Item>> {
  public render() {
    return (
      <FormContext.Consumer>
        {formContext => (
          <FieldGroupContext.Consumer>
            {groupContext => (
              <FieldArrayContext.Consumer>
                {arrayContext => (
                  <FieldArrayItemsWithoutContext
                    {...this.props}
                    formContext={assertNotNull(formContext, 'You cannot use the FieldArrayItems outside a form')}
                    groupContext={groupContext}
                    arrayContext={arrayContext}
                  />
                )}
              </FieldArrayContext.Consumer>
            )}
          </FieldGroupContext.Consumer>
        )}
      </FormContext.Consumer>
    );
  }
}

export interface FieldArrayItemsWithoutContextProps<Item> extends FieldArrayItemsProps<Item> {
  formContext: FormContextValue<any>;
  groupContext: FieldGroupContextValue;
  arrayContext: FieldArrayContextValue;
}

export class FieldArrayItemsWithoutContext<Item = any> extends React.Component<FieldArrayItemsWithoutContextProps<Item>> {
  private array: Item[];

  public render() {
    const { render } = this.props;
    const getKey = this.getKeyFunction();
    this.array = this.getArray();
    return (
      <>
      {this.array.map((item, index) => (
        <FieldGroup
          key={getKey(item, index)}
          name={this.getNameForItem(item, index)}
          accessor={this.getAccessorForItem(index)}
        >
          {render({
            remove: () => this.removeItem(item, index),
            setArray: this.setArray,
            index,
            item
          })}
        </FieldGroup>
      ))}
      </>
    );
  }

  public shouldComponentUpdate(nextProps: FieldArrayItemsWithoutContextProps<any>) {
    return isShallowEqual(this.props, nextProps);
  }

  private getArray(): Item[] {
    const { form: { state: { model }} } = this.props.formContext;
    const { path = '' } = this.props.groupContext;
    return selectDeep({ object: model, path });
  }

  private removeItem(item: Item, index: number): void {
    const { onFieldChange } = this.props.formContext;
    const identifier = this.getItemIdentifier(item, index);
    const path = this.getItemPath(index);
    onFieldChange(identifier, path, DELETE);
  }

  private getItemPath(index: number): Path {
    return createPath(this.props.groupContext.path, index);
  }

  private getItemIdentifier(item: Item, index: number): Path {
    return createPath(this.props.groupContext.namespace, this.getNameForItem(item, index));
  }

  private getNameForItem(item: Item, index: number): string {
    const getKey = this.getKeyFunction();
    return String(getKey(item, index));
  }

  private getKeyFunction(): GetKey<Item> {
    const { getKey } = this.props.arrayContext;
    if (!getKey) {
      throw new Error(`Invalid context. Please make sure that "FieldArrayItems" is wrapped by a "FieldArray"`);
    }
    return getKey;
  }

  private setArray = (newArray: Item[]): void => {
    const { onFieldChange } = this.props.formContext;
    const { path = '', namespace = '' } = this.props.groupContext;
    onFieldChange(namespace, path, newArray);
  };

  private getAccessorForItem(index: number): string {
    return String(index);
  }
}
