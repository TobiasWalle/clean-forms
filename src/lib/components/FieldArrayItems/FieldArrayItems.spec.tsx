import { mount, render } from 'enzyme';
import * as React from 'react';
import { withArrayContext, withFormContext, withGroupContext } from '../../../testUtils/context';
import { InputField } from '../../../testUtils/InputField';
import { mockFormContext } from '../../../testUtils/mockFormContext';
import { FieldArrayContextValue } from '../../contexts/field-array-context';
import { FieldGroupContextValue } from '../../contexts/field-group-context';
import { FormContextValue, OnFieldChange } from '../../contexts/form-context';
import { DELETE } from '../../utils';
import { FieldArrayItems } from './FieldArrayItems';

describe('FieldArrayItems', () => {
  it('should render items', () => {
    const model = {
      array: [
        { a: 0 },
        { a: 1 },
        { a: 2 },
      ]
    };
    const formContext: FormContextValue<typeof model> = mockFormContext(model);
    const groupContext: FieldGroupContextValue = {
      namespace: 'array',
      path: 'array',
    };
    const arrayContext: FieldArrayContextValue = {
      getKey: (_, i) => i
    };
    const element = render(
      withFormContext(formContext)(
        withGroupContext(groupContext)(
          withArrayContext(arrayContext)(
            <FieldArrayItems render={() => <InputField name={'a'}/>}/>
          )
        )
      )
    );

    expect(element).toMatchSnapshot();
  });

  it('should pass remove callback to render function', () => {
    const model = {
      array: [
        { a: 0 },
        { a: 1 },
        { a: 2 },
      ]
    };
    type Model = typeof model;
    const onFieldChange: OnFieldChange<Model> = jest.fn();
    const formContext: FormContextValue<typeof model> = mockFormContext(model, { onFieldChange });
    const groupContext: FieldGroupContextValue = {
      namespace: 'array',
      path: 'array',
    };
    const arrayContext: FieldArrayContextValue = {
      getKey: (_, i) => `key_${i}`
    };
    const element = mount(
      withFormContext(formContext)(
        withGroupContext(groupContext)(
          withArrayContext(arrayContext)(
            <FieldArrayItems
              render={({ remove }) => (
                <div>
                  <InputField name={'a'}/>
                  <button onClick={remove}>Remove</button>
                </div>
              )}/>
          )
        )
      )
    );

    element.find('button').first().simulate('click');

    expect(onFieldChange).toHaveBeenCalledWith('array.key_0', 'array.0', DELETE);
  });

  it('should pass item to render function', () => {
    const item = { a: 0 };
    const model = {
      array: [item]
    };
    type Model = typeof model;
    const onFieldChange: OnFieldChange<Model> = jest.fn();
    const formContext: FormContextValue<typeof model> = mockFormContext(model, { onFieldChange });
    const groupContext: FieldGroupContextValue = {
      namespace: 'array',
      path: 'array',
    };
    const arrayContext: FieldArrayContextValue = {
      getKey: (_, i) => i
    };
    const renderItem = jest.fn(() => <div/>);
    mount(
      withFormContext(formContext)(
        withGroupContext(groupContext)(
          withArrayContext(arrayContext)(
            <FieldArrayItems render={renderItem}/>
          )
        )
      )
    );

    expect(renderItem).toHaveBeenCalledTimes(1);
    expect(renderItem.mock.calls[0][0].item).toBe(item);
  });

  it('should pass index to render function', () => {
    const model = {
      array: [{ a: 0 }, { a: 1 }]
    };
    type Model = typeof model;
    const onFieldChange: OnFieldChange<Model> = jest.fn();
    const formContext: FormContextValue<typeof model> = mockFormContext(model, { onFieldChange });
    const groupContext: FieldGroupContextValue = {
      namespace: 'array',
      path: 'array',
    };
    const arrayContext: FieldArrayContextValue = {
      getKey: (_, i) => i
    };
    const renderItem = jest.fn(() => <div/>);
    mount(
      withFormContext(formContext)(
        withGroupContext(groupContext)(
          withArrayContext(arrayContext)(
            <FieldArrayItems render={renderItem}/>
          )
        )
      )
    );

    expect(renderItem).toHaveBeenCalledTimes(2);
    expect(renderItem.mock.calls[0][0].index).toBe(0);
    expect(renderItem.mock.calls[1][0].index).toBe(1);
  });

  it('should pass setArray callback to render function', () => {
    const model = {
      array: [
        { a: 0 },
        { a: 1 },
        { a: 2 },
      ]
    };
    type Model = typeof model;
    const onFieldChange: OnFieldChange<Model> = jest.fn();
    const formContext: FormContextValue<typeof model> = mockFormContext(model, { onFieldChange });
    const groupContext: FieldGroupContextValue = {
      namespace: 'array',
      path: 'array',
    };
    const arrayContext: FieldArrayContextValue = {
      getKey: (_, i) => i
    };
    const element = mount(
      withFormContext(formContext)(
        withGroupContext(groupContext)(
          withArrayContext(arrayContext)(
            <FieldArrayItems
              render={({ setArray }) => (
                <div>
                  <InputField name={'a'}/>
                  <button onClick={() => setArray([])}>Remove</button>
                </div>
              )}/>
          )
        )
      )
    );

    element.find('button').first().simulate('click');

    expect(onFieldChange).toHaveBeenCalledWith('array', 'array', []);
  });
});
