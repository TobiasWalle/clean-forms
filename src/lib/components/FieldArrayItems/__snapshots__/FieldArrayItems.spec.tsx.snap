// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldArrayItems should render items 1`] = `
Array [
  <input
    name="a"
    value="0"
  />,
  <input
    name="a"
    value="1"
  />,
  <input
    name="a"
    value="2"
  />,
]
`;
